examples/cryptocurrency-advanced/backend/src/schema.rs

```diff
@@ -14,94 +14,106 @@
 
 //! Cryptocurrency database schema.
 
-use exonum_merkledb::{IndexAccess, ObjectHash, ProofListIndex, ProofMapIndex};
+use exonum_merkledb::{AccessExt, IndexAccessMut, ObjectHash, ProofListIndex, ProofMapIndex};
 
 use exonum::crypto::{Hash, PublicKey};
 
 use crate::{wallet::Wallet, INITIAL_BALANCE};
 
+const NOT_INITIALIZED: &str = "Cryptocurrency schema is not initialized";
+
 /// Database schema for the cryptocurrency.
 #[derive(Debug)]
-pub struct Schema<'a, T> {
-    service_name: &'a str,
+pub struct Schema<T: AccessExt> {
+    /// Map of wallet keys to information about the corresponding account.
+    pub wallets: ProofMapIndex<T::Base, PublicKey, Wallet>,
     access: T,
 }
 
-impl<'a, T> Schema<'a, T>
-where
-    T: IndexAccess,
-{
+impl<T: AccessExt> Schema<T> {
     /// Creates a new schema from the database view.
-    pub fn new(service_name: &'a str, access: T) -> Self {
-        Schema {
+    pub fn new(access: T) -> Self {
+        Self {
+            wallets: access.proof_map("wallets").expect(NOT_INITIALIZED),
             access,
-            service_name,
         }
     }
 
-    /// Returns `ProofMapIndex` with wallets.
-    pub fn wallets(&self) -> ProofMapIndex<T, PublicKey, Wallet> {
-        ProofMapIndex::new(
-            [self.service_name, ".wallets"].concat(),
-            self.access.clone(),
-        )
-    }
-
     /// Returns history of the wallet with the given public key.
-    pub fn wallet_history(&self, public_key: &PublicKey) -> ProofListIndex<T, Hash> {
-        ProofListIndex::new_in_family(
-            [self.service_name, ".wallet_history"].concat(),
-            public_key,
-            self.access.clone(),
-        )
+    pub fn wallet_history(&self, public_key: &PublicKey) -> Option<ProofListIndex<T::Base, Hash>> {
+        self.access.proof_list(("wallet_history", public_key))
     }
 
     /// Returns wallet for the given public key.
     pub fn wallet(&self, pub_key: &PublicKey) -> Option<Wallet> {
-        self.wallets().get(pub_key)
+        self.wallets.get(pub_key)
     }
 
     /// Returns the state hash of cryptocurrency service.
     pub fn state_hash(&self) -> Vec<Hash> {
-        vec![self.wallets().object_hash()]
+        vec![self.wallets.object_hash()]
+    }
+}
+
+impl<T> Schema<T>
+where
+    T: AccessExt,
```

Comment:
> was not really good

What's the reasoning for it? Note that there is a single method (or field) to access an index from the schema, which was no the case in the previous differentiated iteration.

`ensure_*` methods are only required during initialization, and so can be called exactly once for singular indexes from a single place (`Schema::initialize()`). For indexes in index families the situation is admittedly worse, but I don't think it could be made better without reverting to the old concept of index families (i.e., the entire family is constructed at the same time).